Mathlib/Analysis/NormedSpace/HomeomorphBall.lean

```diff
@@ -0,0 +1,151 @@
+    have : 0 < 1 - ‖y‖ ^ 2 := by nlinarith [norm_nonneg y, mem_ball_zero_iff.1 hy]
+    field_simp [norm_smul, smul_smul, this.ne', sq_abs, Real.sq_sqrt this.le,
+      ← Real.sqrt_div this.le]
+  open_source := isOpen_univ
+  open_target := isOpen_ball
+  continuous_toFun := by
+    suffices Continuous fun (x:E) => (1 + ‖x‖ ^ 2).sqrt⁻¹
+     from (this.smul continuous_id).continuousOn
+    refine' Continuous.inv₀ _ fun x => Real.sqrt_ne_zero'.mpr (by positivity)
+    continuity
+  continuous_invFun := by
+    have : ∀ y ∈ ball (0 : E) 1, (1 - ‖(y : E)‖ ^ 2).sqrt ≠ 0 := fun y hy ↦ by
+      rw [Real.sqrt_ne_zero']
+      nlinarith [norm_nonneg y, mem_ball_zero_iff.1 hy]
+    exact ContinuousOn.smul (ContinuousOn.inv₀
+      (continuousOn_const.sub (continuous_norm.continuousOn.pow _)).sqrt this) continuousOn_id
+
+@[simp]
+theorem LocalHomeomorph.univUnitBall_apply_zero : univUnitBall (0 : E) = 0 := by
+  simp [LocalHomeomorph.univUnitBall_apply]
+
+@[simp]
+theorem LocalHomeomorph.univUnitBall_symm_apply_zero : univUnitBall.symm (0 : E) = 0 := by
+  simp [LocalHomeomorph.univUnitBall_symm_apply]
+
+/-- A (semi) normed real vector space is homeomorphic to the unit ball in the same space.
+This homeomorphism sends `x : E` to `(1 + ‖x‖²)^(- ½) • x`.
+
+In many cases the actual implementation is not important, so we don't mark the projection lemmas
+`Homeomorph.unitBall_apply_coe` and `Homeomorph.unitBall_symm_apply` as `@[simp]`.
+
+See also `Homeomorph.contDiff_unitBall` and `LocalHomeomorph.contDiffOn_unitBall_symm`
+for smoothness properties that hold when `E` is an inner-product space. -/
+@[simps! (config := .lemmasOnly)]
+def Homeomorph.unitBall : E ≃ₜ ball (0 : E) 1 :=
+  (Homeomorph.Set.univ _).symm.trans LocalHomeomorph.univUnitBall.toHomeomorphSourceTarget
+#align homeomorph_unit_ball Homeomorph.unitBall
+
+@[simp]
+theorem Homeomorph.coe_unitBall_apply_zero :
+    (Homeomorph.unitBall (0 : E) : E) = 0 :=
+  LocalHomeomorph.univUnitBall_apply_zero
+#align coe_homeomorph_unit_ball_apply_zero Homeomorph.coe_unitBall_apply_zero
+
+variable {P : Type _} [PseudoMetricSpace P] [NormedAddTorsor E P]
+
+namespace LocalHomeomorph
+
+/-- Affine homeomorphism `(r • · +ᵥ c)` between a normed space and an add torsor over this space,
+interpreted as a `LocalHomeomorph` between `Metric.ball 0 1` and `Metric.ball c r`. -/
+@[simps!]
+def unitBallBall (c : P) (r : ℝ) (hr : 0 < r) : LocalHomeomorph E P :=
+  ((Homeomorph.smulOfNeZero r hr.ne').trans
+      (IsometryEquiv.vaddConst c).toHomeomorph).toLocalHomeomorphOfImageEq
+      (ball 0 1) isOpen_ball (ball c r) <| by
+    change (IsometryEquiv.vaddConst c) ∘ (r • ·) '' ball (0 : E) 1 = ball c r
+    rw [image_comp, image_smul, smul_unitBall hr.ne', IsometryEquiv.image_ball]
+    simp [abs_of_pos hr]
+
+/-- If `r > 0`, then `LocalHomeomorph.univBall c r` is a smooth local homeomorphism
+with `source = Set.univ` and `target = Metric.ball c r`.
+Otherwise, it is the translation by `c`. -/
```

Comment:
I think it's worth highlighting this (very nice) feature explicitly.
```suggestion
Otherwise, it is the translation by `c`. (Thus in all cases, it sends `0` to `c`, see
`LocalHomeomorph.univBall_apply_zero`.) -/
```